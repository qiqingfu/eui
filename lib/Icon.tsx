import React from "react"
import "./icons"
import "./styles/icon.scss"

interface IconProps {
	name: string
}

const Icon: React.FunctionComponent<IconProps> = (props) => {
	return (
		// https://stackoverflow.com/questions/26815738/svg-use-tag-and-reactjs
		<svg className="e-ui-icon">
			<use xlinkHref={`#${props.name}`} />
		</svg>
	)
}

export default Icon