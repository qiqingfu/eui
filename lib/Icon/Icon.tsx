import React from 'react';
import classNames from 'classnames';
import '../../icons';
import './icon.scss';

export interface IconProps<T = SVGElement> extends React.SVGAttributes<T> {
  name: string;
  color?: string;
  size?: number;
}

const Icon: React.FunctionComponent<IconProps> = ({ name, className, ...props }: IconProps) => {
  const { style, restProps } = useStyle(props as IconProps);

  return (
    // https://stackoverflow.com/questions/26815738/svg-use-tag-and-reactjs
    <svg className={classNames('e-ui-icon', className)} {...restProps} style={style}>
      <use xlinkHref={`#${name}`} />
    </svg>
  );
};

function useStyle(props: IconProps) {
  const { style = {}, color, size, ...restProps } = props;

  if (color) style.fill = color;
  if (size) style.fontSize = size;

  return {
    style,
    restProps,
  };
}

export default Icon;
