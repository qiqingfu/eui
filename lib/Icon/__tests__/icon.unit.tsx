import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Icon from "../index";

describe("Icon Component", () => {
  it("Icon Render", () => {
    const tree = renderer.create(<Icon name="delete" />);
    expect(tree).toMatchSnapshot();
  });

  it("Icon Props.name", () => {
    const wrapper = shallow(<Icon name="delete" />);
    const result = wrapper.find("use").props();
    expect(result.xlinkHref).toBe("#delete");
  });

  it("Icon Props.color", () => {
    const wrapper = shallow(<Icon name="delete" color="#f00" />);
    const result = wrapper.find(".e-ui-icon").props();
    expect(result.style?.fill).toBe("#f00");
  });

  it("Icon Props.size", () => {
    const wrapper = shallow(<Icon name="delete" size={40} />);
    const result = wrapper.find(".e-ui-icon").props();
    expect(result.style?.fontSize).toBe(40);
  });

  it("Icon click", () => {
    const fn = jest.fn();
    const wrapper = shallow(<Icon name="delete" onClick={fn} />);
    wrapper.simulate("click");
    expect(fn).toBeCalledTimes(1);
  });
});
