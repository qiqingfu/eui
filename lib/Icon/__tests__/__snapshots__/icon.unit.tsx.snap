// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component Icon Render 1`] = `
<svg
  className="e-ui-icon"
  style={Object {}}
>
  <use
    xlinkHref="#delete"
  />
</svg>
`;
